pallets/attestation/src/migrations.rs

```diff
@@ -0,0 +1,133 @@
+{
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		log::info!("Attestation: Initiating migration");
+
+		let onchain_storage_version = Pallet::<T>::on_chain_storage_version();
+		if onchain_storage_version.eq(&StorageVersion::new(1)) {
+			StorageVersion::new(2).put::<Pallet<T>>();
+			return do_migration::<T>();
+		}
+
+		log::info!(
+			"Attestation: No migration needed. This file should be deleted. Current storage version: {:?}, Required Version for update: {:?}", 
+			onchain_storage_version,
+			StorageVersion::new(1)
+		);
+
+		<T as frame_system::Config>::DbWeight::get().reads_writes(0, 0)
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<sp_std::vec::Vec<u8>, &'static str> {
+		use sp_std::vec;
+
+		let has_all_user_no_holds = Attestations::<T>::iter_values()
+			.map(|details: crate::AttestationDetails<T>| {
+				kilt_support::migration::has_user_reserved_balance::<AccountIdOf<T>, CurrencyOf<T>>(
+					&details.deposit.owner,
+					&T::RuntimeHoldReason::from(HoldReason::Deposit),
+				)
+			})
+			.all(|user| user);
+
+		assert!(
+			has_all_user_no_holds,
+			"Pre Upgrade Attestation: there are users with holds!"
+		);
+
+		assert_eq!(Pallet::<T>::on_chain_storage_version(), StorageVersion::new(1));
+
+		log::info!("Attestation: Pre migration checks successful");
+
+		Ok(vec![])
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(_pre_state: sp_std::vec::Vec<u8>) -> Result<(), &'static str> {
+		use frame_support::traits::fungible::InspectHold;
+
+		Attestations::<T>::iter().try_for_each(|(key, details)| -> Result<(), &'static str> {
+			let hold_balance: u128 = <T as Config>::Currency::balance_on_hold(
+				&T::RuntimeHoldReason::from(HoldReason::Deposit),
+				&details.deposit.owner,
+			)
+			.saturated_into();
+			assert!(
+				details.deposit.amount.saturated_into::<u128>() <= hold_balance,
+				"Attestation: Hold balance is not matching for attestation {:?}. Expected hold: {:?}. Real hold: {:?}",
+				key,
+				details.deposit.amount,
+				hold_balance
+			);
+
+			Ok(())
+		})?;
+
+		assert_eq!(Pallet::<T>::on_chain_storage_version(), StorageVersion::new(2));
+
+		log::info!("Attestation: Post migration checks successful");
+		Ok(())
+	}
+}
+
+fn do_migration<T: Config>() -> Weight
+where
+	<T as Config>::Currency: ReservableCurrency<T::AccountId>,
+{
+	Attestations::<T>::iter()
+		.map(|(key, attestations_detail)| -> Weight {
+			let deposit = attestations_detail.deposit;
+			let error = switch_reserved_to_hold::<AccountIdOf<T>, CurrencyOf<T>>(
+				deposit.owner,
+				&T::RuntimeHoldReason::from(HoldReason::Deposit),
+				deposit.amount.saturated_into(),
+			);
+
+			if error.is_ok() {
+				return <T as frame_system::Config>::DbWeight::get().reads_writes(1, 1);
+			}
+
+			log::error!(
+				" Attestation: Could not convert reserves to hold from attestation: {:?} error: {:?}",
+				key,
+				error
+			);
+
+			<T as frame_system::Config>::DbWeight::get().reads_writes(0, 0)
```

Comment:
How can the weight of this be `(0, 0)`? At the very least we need two storage reads and one write, to read and delete the old storage and write the new one. The difference between the `ok` case and the error case is just a `log::error!`, so the weight is actually the same.

Reply:
Resolved: [e5966d6](https://github.com/KILTprotocol/kilt-node/pull/527/commits/e5966d6b775b4b648f609ba3b0af6e77973f8e45)